server/src/storage/mod.rs

```diff
@@ -314,6 +355,48 @@ impl StorageEngine {
     }
 }
 
+/// Log used to record changes in specific attributes in the storage engine.
+pub struct Log {
+    /// Builder consisting of millisecond precision timestamps.
+    timestamps: TimestampBuilder,
+    /// Builder consisting of values.
+    values: UInt32Builder,
+    /// Since the values builder is cleared when the log is finished, the last value is saved separately.
+    last_value: isize,
+}
+
+impl Log {
+    fn new() -> Self {
+        Self {
+            timestamps: TimestampBuilder::new(),
+            values: UInt32Builder::new(),
+            last_value: 0,
+        }
+    }
+
+    /// Add a new entry to the log, where the timestamp is the current milliseconds since the Unix
+    /// epoch and the value is either set directly or based on the last value in the log.
+    fn add_entry(&mut self, value: isize, based_on_last: bool) {
```

Comment:
Maybe rename to `append()` or `append_entry()`. Also maybe replace the `bool` with one method for `based_on_last: false` and another for `based_on_last: true` that both calls a shared method with most of the logic. Although since all three methods would have the same level of visibility, it may be more confusion than the current API.

Reply:
Renamed the method to `append()`. I don't mind the other change but as you mentioned, it might be more confusing so I have kept the current API.